resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -144,13 +144,94 @@ public void operationComplete(Future<T> future) throws Exception {
     }
 
     private void internalResolve(Promise<T> promise) {
-        InetSocketAddress nameServerAddrToTry = nameServerAddrs.next();
+        DnsServerAddressStream nsAddrs = getCachedDnsServers(hostname);
+        if (nsAddrs == null) {
+            // Nothing found in the cache just start from the nameServerAddrs.
+            nsAddrs = nameServerAddrs;
+        }
         for (DnsRecordType type: parent.resolveRecordTypes()) {
-            query(nameServerAddrToTry, new DefaultDnsQuestion(hostname, type), promise);
+            query(nsAddrs, new DefaultDnsQuestion(hostname, type), promise);
         }
     }
 
-    private void query(InetSocketAddress nameServerAddr, final DnsQuestion question, final Promise<T> promise) {
+    /**
+     * Add a DNS server to the cache if its not a root server.
+     */
+    private void addCacheDnsServer(DnsServerNames.ServerName name, InetAddress resolved, long ttl) {
+        if (!name.isRootServer()) {
+            // Cache NS record if not for a root server as we should never cache for root servers.
+            parent.nsCache().cache(name.delegateName(),
+                    additionals, resolved, ttl, parent.ch.eventLoop());
+        }
+    }
+
+    /**
+     * Returns the {@link DnsServerAddressStream} that was cached for the given hostname or {@code null} if non
+     *  could be found.
+     */
+    private DnsServerAddressStream getCachedDnsServers(String hostname) {
+        int len = hostname.length();
+
+        if (len == 0 || len == 1 && hostname.charAt(0) == '.') {
+            // We never cache for root servers.
+            return null;
+        }
+
+        // We always store in the case with a trailing '.'.
+        if (hostname.charAt(len - 1) != '.') {
+            hostname += ".";
+        }
+
+        // We start from the closed match and then move down.
+        for (;;) {
+            int idx = hostname.indexOf('.');
+            if (idx <= 0) {
+                // We are not interested in handling '.' as we should never serve the root servers from cache.
+                return null;
+            }
+            // Skip '.' as well.
+            hostname = hostname.substring(idx + 1);
```

Comment:
@fenik17 at least thats what I wanted to do with the method... return stuff from the cache that is not the "root" domain. Please note that we also not add it to the cache if its the root domain.